### PR TITLE
Remove MS Remote Desktop support

### LINKED ISSUE #1159: Drop Microsoft Remote Desktop
## :strawberry: Desired behaviour

We should strongly consider dropping Microsoft Remote Desktop for the following reasons:

1. Recent penetration tests show more issues in Microsoft Remote Desktop than Guacamole. Dropping Microsoft Remote Desktop would close the following issues:
  - alan-turing-institute/data-safe-haven#361 
  - alan-turing-institute/data-safe-haven-production#45 
  - alan-turing-institute/data-safe-haven-production#46 
  - alan-turing-institute/data-safe-haven-production#49 
  - alan-turing-institute/data-safe-haven-production#62 
  - alan-turing-institute/data-safe-haven-production#59 
  - alan-turing-institute/data-safe-haven-production#54 
2. Microsoft prefer Guacamole [in their own TRE](https://github.com/microsoft/AzureTRE/)
3. Dropping support would allow us to remove three VMs: `RDG-SRE-{SRE-ID}`; `APP-SRE-{SRE-ID}`; `NPS-SHM-{SHM-ID}` and their associated setup and code
4. Only supporting one remote desktop would reduce our support surface
5. This would also mean that we could tighten our firewall rules allowing connections to various Windows update services to only apply to SHM VMs.

Thoughts welcome @martintoreilly @JimMadge

### PR DESCRIPTION
**Pull request by craddm**
## :white_check_mark: Checklist

- [x] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [x] You are targeting the appropriate branch. If you're not certain which one this is, it should be **`develop`**.
- [x] Your branch is up-to-date with the **target branch** (it probably was when you started, but it may have changed since then).
- [x] You have marked this pull request as a **draft** and added `'[WIP]'` to the title if needed (if you're not yet ready to merge).
- [x] You have formatted your code using appropriate automated tools (for example `./tests/AutoFormat_Powershell.ps1 -TargetPath <path to file or directory>` for Powershell).

### :arrow_heading_up: Summary

Removes support for MS Remote Desktop. 

- [x] NPS removed from SHM deployment
- [x] Code relating to MSRDS removed from SRE deployment scripts
- [x] MSRDS related documentation removed

NOTE: **do not merge before Release 4.0.4 is complete**

### :closed_umbrella: Related issues

Closes #1159 

### :microscope: Tests

Deployed an SHM and Guacamole Tier 2 SRE successfully. Rebuilt docs locally.

### CODE REVIEW

**Comment by JimMadge** [comment]
@craddm Is this ready to review?

We can absolutely merge this into develop, just not into the release branch. (@jemrobinson feel free to disagree).

**Comment by jemrobinson** [comment]
If we're confident that the release branch is good-to-go or close to it then yes. Thoughts @edwardchalstrey1 ?

**Review by jemrobinson — APPROVED** [review]
Great work here @craddm. It's amazing how many interconnected pieces were just there to support MSRDS!

**Comment by JimMadge** [comment]
I would say we should take the position that people are free to contribute to develop without worrying about release branches.

If changes on develop should also be on the release branch, they can be cherry-picked.

**Comment by craddm** [comment]
Minor thing that I've just realised - the documentation images still include MSRDS specific things (I'm looking at you, `Network Policy Server`), so they'll need updating too. Are there master versions somewhere? That said, probably need to do a pass renaming Active Directory to Entra at some point in the near future, so maybe do that and the images separately?

**Comment by jemrobinson** [comment]
> Minor thing that I've just realised - the documentation images still include MSRDS specific things (I'm looking at you, `Network Policy Server`), so they'll need updating too. Are there master versions somewhere?

No. Are these diagrams (which we can re-draw) or screenshots (which we can re-take)?

> That said, probably need to do a pass renaming Active Directory to Entra at some point in the near future, so maybe do that and the images separately?

What is Entra?

**Comment by edwardchalstrey1** [comment]
> If we're confident that the release branch is good-to-go or close to it then yes. Thoughts @edwardchalstrey1 ?

If this PR makes changes to SHM scripts should a test deployment be carried out first? I'd suggest to not merge into `4.0.4` but make a new version with this plus any other recent additions on the develop branch.

Since we are removing functionality here, should we make a `4.1.0` release? Looking at https://github.com/alan-turing-institute/data-safe-haven/blob/develop/VERSIONING.md#numbering it's not totally clear based on our current guidelines, but feels like removing functionality is more than a patch? I'm not sure though

**Comment by jemrobinson** [comment]
This PR can't go into v4.0.4. We need at least one release that says "MSRDS is deprecated" before one that removes MSRDS. It can go into `develop` though.

Our versioning says:

```
MAJOR: Using this release together with an existing environment will require redeployment or patching of some or all existing components. An SRE will not be compatible with an SHM (or another SRE) deployed from a different MAJOR version.
MINOR: Adds new functionality, but new SREs can be deployed into existing environments deployed from any version sharing the same MAJOR number without these existing SHM and SRE elements needing to be redeployed or patched.
PATCH: Fixes bugs only and new SREs can be deployed into existing environments deployed from any version sharing the same MAJOR number without these existing SHM and SRE elements needing to be redeployed or patched.
```

so I think this wouldn't be a MAJOR bump, but might be a MINOR one.

**Comment by craddm** [comment]
> No. Are these diagrams (which we can re-draw) or screenshots (which we can re-take)?

Diagrams like this one - 
https://data-safe-haven.readthedocs.io/en/latest/_images/shm_architecture.png
 
> > That said, probably need to do a pass renaming Active Directory to Entra at some point in the near future, so maybe do that and the images separately?
> 
> What is Entra?

New name for Azure Active Directory - https://www.microsoft.com/en-gb/security/business/identity-access/azure-active-directory

**Comment by jemrobinson** [comment]
> Diagrams like this one - https://data-safe-haven.readthedocs.io/en/latest/_images/shm_architecture.png

Ah right. Yes, those ones *do* have a canonical source. They have draw.io XML files in Sharepoint under `design > 2021-10 implementation_paper > xml`

> > What is Entra?
> 
> New name for Azure Active Directory - https://www.microsoft.com/en-gb/security/business/identity-access/azure-active-directory

Urgh, thanks Microsoft. I see another API/function name/variable name refactoring on the horizon.

**Comment by craddm** [comment]
Updated the figures and removed some more vestigial network policy server bits from the documentation

**Comment by JimMadge** [comment]
> Updated the figures and removed some more vestigial network policy server bits from the documentation

@craddm Is this ready to go then?

**Comment by jemrobinson** [comment]
Can you resolve the conflicts @craddm ?

**Review by JimMadge — APPROVED** [review]
LGTM this has cleared out a lot of code to support :tada:.

**Comment by JimMadge** [comment]
A few references to MSRDS remaining in the [smoke tests README](https://github.com/alan-turing-institute/data-safe-haven/blob/5c3e9cb1d4d04d9032b211cc6e6509be7593a5b2/tests/srd_smoke_tests/README.md) but that isn't a big deal, it is just an example SRE name.

**Comment by craddm** [comment]
> A few references to MSRDS remaining in the [smoke tests README](https://github.com/alan-turing-institute/data-safe-haven/blob/5c3e9cb1d4d04d9032b211cc6e6509be7593a5b2/tests/srd_smoke_tests/README.md) but that isn't a big deal, it is just an example SRE name.

I found a few more in the security checklist, which I'm just fixing along with the linting issues, so I'll fix this one too.

**Comment by JimMadge** [comment]
```
❯ rg -i 'microsoft remote desktop'
docs/source/deployment/deploy_shm.md
40:- `Microsoft Remote Desktop`

docs/source/deployment/snippets/01_prerequisites.partial.md
19:- `Microsoft Remote Desktop`

docs/source/deployment/snippets/00_symbols.partial.md
23:- This indicates a command which you will need to run remotely on an Azure virtual machine (VM) using `Microsoft Remote Desktop`
24:- Open `Microsoft Remote Desktop` and click `Add Desktop` / `Add PC`
```

Oh, these are fine. This is the desktop app for connecting to remote windows machines.

**Comment by craddm** [comment]
> ```
> ❯ rg -i 'microsoft remote desktop'
> docs/source/deployment/deploy_shm.md
> 40:- `Microsoft Remote Desktop`
> 
> docs/source/deployment/snippets/01_prerequisites.partial.md
> 19:- `Microsoft Remote Desktop`
> 
> docs/source/deployment/snippets/00_symbols.partial.md
> 23:- This indicates a command which you will need to run remotely on an Azure virtual machine (VM) using `Microsoft Remote Desktop`
> 24:- Open `Microsoft Remote Desktop` and click `Add Desktop` / `Add PC`
> ```

It's still necessary to use MS Remote Desktop to interact with the domain controllers

**Comment by JimMadge** [comment]
```
❯ rg -i 'microsoft rds'
deployment/common/Configuration.psm1
857:    # Remote desktop either through Apache Guacamole or Microsoft RDS
```

**Comment by JimMadge** [comment]
```
❯ rg 'RDS'
deployment/secure_research_environment/network_rules/sre-nsg-rules-guacamole.json
17:        "description": "Allow inbound http(s) connections from clients to RDS server. Note that http requests will be upgraded.",

deployment/secure_research_environment/setup/Apply_SRE_Network_Configuration.ps1
41:    # RDS gateway

deployment/administration/SRE_Manage_VMs.ps1
30:        # May be able to simplify this further now that MSRDS is removed
33:        # Start all other VMs before RDS VMs so all services will be available when users can login via RDS

deployment/common/Configuration.psm1
508:        rg               = $shmConfigBase.dnsRecords.resourceGroupName ? $shmConfigBase.dnsRecords.resourceGroupName : "$($shm.rgPrefix)_DNS_RECORDS".ToUpper()
754:                sreArtifactsRDS = "sre-artifacts-rds"
755:                sreScriptsRDS   = "sre-scripts-rds"
857:    # Remote desktop either through Apache Guacamole or Microsoft RDS

deployment/safe_haven_management_environment/desired_state_configuration/dc1Artifacts/source/{742211F9-1482-4D06-A8DE-BA66101933EB}/gpreport.xml
173:    <SOMName>DSGroup2 RDS Session Servers</SOMName>
174:    <SOMPath>DSGroup2.co.uk/DSGroup2 RDS Session Servers</SOMPath>

deployment/safe_haven_management_environment/desired_state_configuration/dc1Artifacts/source/{0AF343A0-248D-4CA5-B19E-5FA46DAE9F9C}/gpreport.xml
165:    <SOMName>DSGroup2 RDS Session Servers</SOMName>
166:    <SOMPath>DSGroup2.co.uk/DSGroup2 RDS Session Servers</SOMPath>

docs/source/deployment/snippets/02_configuration.partial.md
25:  "inboundAccessFrom": "A comma-separated string of IP ranges (addresses or CIDR ranges) from which access to the RDS webclient is permitted. See tip default below for suggestion on how to set this.",
31:  "remoteDesktopProvider": "Which remote desktop provider to use. Either 'ApacheGuacamole' (recommended, tiers 0-3) or 'MicrosoftRDS' (tiers 2-3 only)",

tests/benchmark_shm.json
323:                "Name": "RADIUS_Authenitcation_RDS_to_NPS",
837:                "Name": "RADIUS_Authenitcation_RDS_to_NPS",
1351:                "Name": "RADIUS_Authenitcation_RDS_to_NPS",

tests/benchmark_sre.json
771:                "Name": "RADIUS_Authentication_RDS_to_NPS",

tests/resources/sre_bluet1guac_full_config.json
1309:          "sreArtifactsRDS": "sre-artifacts-rds",
1310:          "sreScriptsRDS": "sre-scripts-rds"

tests/resources/sre_greent2guac_full_config.json
1362:          "sreArtifactsRDS": "sre-artifacts-rds",
1363:          "sreScriptsRDS": "sre-scripts-rds"

tests/resources/sre_bluet3guac_full_config.json
1309:          "sreArtifactsRDS": "sre-artifacts-rds",
1310:          "sreScriptsRDS": "sre-scripts-rds"
```

This one is worth changing

https://github.com/craddm/data-safe-haven/blob/562f9a3bc343af4533e855625b033148892c0b70/docs/source/deployment/snippets/02_configuration.partial.md?plain=1#L31

Another reference to "RDS" in the same file.

I'm not sure what the JSON files in the test directory or the DC desire state files mean. @jemrobinson?

**Comment by craddm** [comment]
> ```
> ❯ rg 'RDS'
> deployment/secure_research_environment/network_rules/sre-nsg-rules-guacamole.json
> 17:        "description": "Allow inbound http(s) connections from clients to RDS server. Note that http requests will be upgraded.",
> 
> deployment/secure_research_environment/setup/Apply_SRE_Network_Configuration.ps1
> 41:    # RDS gateway
> 
> deployment/administration/SRE_Manage_VMs.ps1
> 30:        # May be able to simplify this further now that MSRDS is removed
> 33:        # Start all other VMs before RDS VMs so all services will be available when users can login via RDS
> 
> deployment/common/Configuration.psm1
> 508:        rg               = $shmConfigBase.dnsRecords.resourceGroupName ? $shmConfigBase.dnsRecords.resourceGroupName : "$($shm.rgPrefix)_DNS_RECORDS".ToUpper()
> 754:                sreArtifactsRDS = "sre-artifacts-rds"
> 755:                sreScriptsRDS   = "sre-scripts-rds"
> 857:    # Remote desktop either through Apache Guacamole or Microsoft RDS
> 
> deployment/safe_haven_management_environment/desired_state_configuration/dc1Artifacts/source/{742211F9-1482-4D06-A8DE-BA66101933EB}/gpreport.xml
> 173:    <SOMName>DSGroup2 RDS Session Servers</SOMName>
> 174:    <SOMPath>DSGroup2.co.uk/DSGroup2 RDS Session Servers</SOMPath>
> 
> deployment/safe_haven_management_environment/desired_state_configuration/dc1Artifacts/source/{0AF343A0-248D-4CA5-B19E-5FA46DAE9F9C}/gpreport.xml
> 165:    <SOMName>DSGroup2 RDS Session Servers</SOMName>
> 166:    <SOMPath>DSGroup2.co.uk/DSGroup2 RDS Session Servers</SOMPath>
> 
> docs/source/deployment/snippets/02_configuration.partial.md
> 25:  "inboundAccessFrom": "A comma-separated string of IP ranges (addresses or CIDR ranges) from which access to the RDS webclient is permitted. See tip default below for suggestion on how to set this.",
> 31:  "remoteDesktopProvider": "Which remote desktop provider to use. Either 'ApacheGuacamole' (recommended, tiers 0-3) or 'MicrosoftRDS' (tiers 2-3 only)",
> 
> tests/benchmark_shm.json
> 323:                "Name": "RADIUS_Authenitcation_RDS_to_NPS",
> 837:                "Name": "RADIUS_Authenitcation_RDS_to_NPS",
> 1351:                "Name": "RADIUS_Authenitcation_RDS_to_NPS",
> 
> tests/benchmark_sre.json
> 771:                "Name": "RADIUS_Authentication_RDS_to_NPS",
> 
> tests/resources/sre_bluet1guac_full_config.json
> 1309:          "sreArtifactsRDS": "sre-artifacts-rds",
> 1310:          "sreScriptsRDS": "sre-scripts-rds"
> 
> tests/resources/sre_greent2guac_full_config.json
> 1362:          "sreArtifactsRDS": "sre-artifacts-rds",
> 1363:          "sreScriptsRDS": "sre-scripts-rds"
> 
> tests/resources/sre_bluet3guac_full_config.json
> 1309:          "sreArtifactsRDS": "sre-artifacts-rds",
> 1310:          "sreScriptsRDS": "sre-scripts-rds"
> ```
> 
> This one is worth changing
> 
> https://github.com/craddm/data-safe-haven/blob/562f9a3bc343af4533e855625b033148892c0b70/docs/source/deployment/snippets/02_configuration.partial.md?plain=1#L31
> 
> Another reference to "RDS" in the same file.
> 
> I'm not sure what the JSON files in the test directory or the DC desire state files mean. @jemrobinson?

Hm, with that one - the `remoteDesktopProvider` field in the config file - is it better to remove that completely from the example and provided configs? I'd still maintain a check of its value in the actual deployment scripts, so that anyone trying to run an old MSRDS config script on this version would be given a useful error message, but could otherwise set it up to set `remoteDesktopProvider` to `ApacheGuacamole` if it is not provided in the config.

**Comment by JimMadge** [comment]
> Hm, with that one - the remoteDesktopProvider field in the config file - is it better to remove that completely from the example and provided configs? I'd still maintain a check of its value in the actual deployment scripts, so that anyone trying to run an old MSRDS config script on this version would be given a useful error message, but could otherwise set it up to set remoteDesktopProvider to ApacheGuacamole if it is not provided in the config.

If I understand, yes I think we can keep the variable _and_ if we have a variable we should check that it is valid.

Removing all of this is probably unnecessary work. Also, like you said, it could be useful to prompt people to 'upgrade' and give a useful error message. We can make Guacamole the default though, so that you don't need to specify this in configuration files.

**Comment by jemrobinson** [comment]
I suggest making it a non-configurable option (ie. don't check for in in Configuration.psm1 but just set it to "ApacheGuacamole" there). This would remove it from the example configs but leave it in the full configs.

File: deployment/common/Configuration.psm1 [hunk 1]
```diff
@@ -624,22 +612,15 @@ function Get-SreConfig {
     # Import minimal management config parameters from JSON config file - we can derive the rest from these
     $sreConfigBase = Get-CoreConfig -shmId $shmId -sreId $sreId
 
+    # "MicrosoftRDS" is deprecated and the "remotedDesktopProvider" field now defaults to "ApacheGuacamole"
```
Comment (JimMadge) [inline deployment/common/Configuration.psm1]:
```suggestion
    # Support for "MicrosoftRDS" has been removed. The "remotedDesktopProvider" field now defaults to "ApacheGuacamole"
```

File: deployment/common/Configuration.psm1 [hunk 2]
```diff
@@ -854,8 +860,8 @@ function Get-SreConfig {
         }
     }
 
-    # Remote desktop either through Apache Guacamole or Microsoft RDS
-    # ---------------------------------------------------------------
+    # Remote desktop either through Apache Guacamole
+    # ----------------------------------------------
```
Comment (JimMadge) [inline deployment/common/Configuration.psm1]:
```suggestion
    # Apache Guacamole remote desktop
    # -------------------------------
```

File: deployment/common/Configuration.psm1 [hunk 3]
```diff
@@ -612,6 +612,12 @@ function Get-SreConfig {
     # Import minimal management config parameters from JSON config file - we can derive the rest from these
     $sreConfigBase = Get-CoreConfig -shmId $shmId -sreId $sreId
 
+    # "MicrosoftRDS" is deprecated and the "remotedDesktopProvider" field now defaults to "ApacheGuacamole"
+    if ($sreConfigBase.remoteDesktopProvider -eq "MicrosoftRDS") {
+        Add-LogMessage -Level Warning "MicrosoftRDS support is deprecated. Apache Guacamole will be used to provide Remote Desktop services."
+    }
+    $sreConfigBase.remoteDesktopProvider = "ApacheGuacamole"
```
Comment (JimMadge) [inline deployment/common/Configuration.psm1]: I don't think this does what we want. A user could put in any invalid except "MicrosoftRDS" and would get no warning or error. A user can explicitly select "MicrosoftRDS" and the process will continue, deploying something else instead.

```suggestion
    if ($sreConfigBase.remoteDesktopProvider -ne "ApacheGuacamole") {
        Add-LogMessage -Level Fatal "Support for remote desktops other than ApacheGuacamole has been removed"
    } elseif ($sreConfigBase.remoteDesktopProvider -eq "ApacheGuacamole") {
        Add-LogMessage -Level Warning "The remoteDesktopProvider configuration option has been deprecated and will be removed in the future"
    }
    $sreConfigBase.remoteDesktopProvider = "ApacheGuacamole"
```

or, would this work

```powershell
if ($sreConfigBase.remoteDesktopProvider -ne $null) {
        Add-LogMessage -Level Fatal "Support for remote desktops other than ApacheGuacamole has been removed. Remove the remoteDesktopProvider option from your configuration."
}
$sreConfigBase.remoteDesktopProvider = "ApacheGuacamole"
```

**Review by JimMadge — CHANGES_REQUESTED** [review]

**Review by craddm — COMMENTED** [review]
@JimMadge I think all your comments have been addressed now

**Review by JimMadge — APPROVED** [review]

**Pull request merged**